Clear mouse capture when control removed.

grokys opened the pull request:
When a captured control is removed from the visual tree, clear the mouse capture.

Note that our capture logic needs to be improved - it shouldn't be possible to capture a non-visible or non-enabled control. This doesn't address that but it does fix the crash in #1203.

Also, if a control has been detached from the visual tree but the scene hasn't yet been updated, don't return the control as part of a hit-test.

(Doesn't) Fix #1203

grokys committed:
Clear mouse capture when control removed.

When a captured control is removed from the visual tree, clear the mouse capture.

Note that our capture logic needs to be improved - it shouldn't be possible to capture a non-visible or non-enabled control. This doesn't address that but it does fix the crash in #1203.

Fixes #1203

kekekeks reviewed (approved):
LGTM

Commit:
Merge branch 'master' into fixes/1203-clear-capture-on-control-removed

danwalmsley commented:
Doesn't fix #1203  but still a required fix, LGTM.

danwalmsley reviewed (approved):
LGTM

kekekeks reviewed (changes requested):
It seems that DetachedFromVisualTree is not being called in time or not being called at all.

danwalmsley reviewed (changes requested):
I added some debugging code to see if Capture = null happens, as @kekekeks says, its not getting called.

kekekeks commented:
I see several places in AvalonStudio where `OnDetachedFromVisualTree` is overridden without calling the base class. I think we need to fire `DetachedFromVisualTree` event outside of `OnDetachedFromVisualTree`

grokys committed:
Always raise logical tree attach/detach events.

Previously, if the user overrode `OnAttachedToLogicalTree` or `OnDetachedFromLogicalTree` and forgot to call the base implementation, the events wouldn't be raised. It's very important that these events get raised so moved the code to raise them to `OnAttachedToLogicalTreeCore`/`OnDetachedFromLogicalTreeCore`.

grokys commented:
I've moved raising the events to `OnAttachedToLogicalTreeCore`/`OnDetachedFromLogicalTreeCore` so they should always get raised even if the user overrides OnAttachedToLogicalTree` or `OnDetachedFromLogicalTree` and forgets to call the base implementation.

Commit:
Merge branch 'master' into fixes/1203-clear-capture-on-control-removed

grokys committed:
Always raise visual tree attach/detach events.

Previously, if the user overrode `OnAttachedToVisualTree` or `OnDetachedFromVisualTree` and forgot to call the base implementation, the events wouldn't be raised. It's very important that these events get raised so moved the code to raise them to `OnAttachedToVisualTreeCore`/`OnDetachedFromVisualTreeCore`.

grokys committed:
Don't return detached controls in HitTest.

If a control has been detached from the visual tree but the scene hasn't yet been updated, don't return the control as part of a hit-test.

danwalmsley commented on src/Avalonia.Input/MouseDevice.cs:
```diff
@@ -55,6 +72,7 @@ public Point Position
         /// </remarks>
         public virtual void Capture(IInputElement control)
         {
+            // TODO: Check visibility and enabled state before setting capture.
```
see https://github.com/AvaloniaUI/Avalonia/issues/1211

danwalmsley reviewed (approved):
LGTM and tested with AS

Commit:
Merge branch 'master' into fixes/1203-clear-capture-on-control-removed

Commit:
Merge branch 'master' into fixes/1203-clear-capture-on-control-removed
